File: app/src/main/java/co/garmax/materialflashlight/ui/LightFragment.java
package co.garmax.materialflashlight.ui;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.content.LocalBroadcastManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import co.garmax.materialflashlight.R;
import co.garmax.materialflashlight.features.LightManager;
import co.garmax.materialflashlight.features.SettingsRepository;
import co.garmax.materialflashlight.features.modules.ScreenModule;

public class LightFragment extends BaseFragment {

    @BindView(R.id.fab)
    FloatingActionButton fab;

    @BindView(R.id.layout_root)
    View layoutRoot;

    @Inject
    LightManager lightManager;

    @Inject
    SettingsRepository settingsRepository;

    /**
     * Receive and handle commands from screen module
     */
    private BroadcastReceiver broadcastReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (intent == null) return;

            // Get brightnessObservable value
            int brightness = intent.getIntExtra(ScreenModule.EXTRA_BRIGHTNESS_PERCENT, 100);

            setBrightness(brightness);
        }
    };

    @Override
    boolean isInImmersiveMode() {
        return true;
    }

    @Override
    public void onResume() {
        super.onResume();

        LocalBroadcastManager.getInstance(requireContext()).
                registerReceiver(broadcastReceiver,
                        new IntentFilter(ScreenModule.ACTION_SCREEN_MODULE));
    }

    @Override
    public void onPause() {
        super.onPause();

        LocalBroadcastManager.getInstance(requireContext()).unregisterReceiver(broadcastReceiver);
    }

    @Override
    public void onStop() {
        super.onStop();

        // Turn off light because screen not visible and for this mode it makes no sense
        if(lightManager.isTurnedOn()) {
            lightManager.turnOff();
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_light, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        ButterKnife.bind(this, view);
        fab.setKeepScreenOn(settingsRepository.isKeepScreenOn());
    }

    @OnClick(R.id.fab)
    void onClick() {
        lightManager.turnOff();
    }

    private void setBrightness(int percent) {

        int color = 255 * percent / 100;

        layoutRoot.setBackgroundColor(Color.rgb(color, color, color));
    }
}

File: app/src/main/java/co/garmax/materialflashlight/ui/light/LightFragment.kt
package co.garmax.materialflashlight.ui.light

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import co.garmax.materialflashlight.databinding.FragmentLightBinding
import co.garmax.materialflashlight.features.LightManager
import co.garmax.materialflashlight.features.modules.ScreenModule
import co.garmax.materialflashlight.repositories.SettingsRepository
import co.garmax.materialflashlight.ui.BaseFragment
import org.koin.android.ext.android.inject

/**
 * Light simulation screen
 */
class LightFragment : BaseFragment() {

    override val isInImmersiveMode = true

    private val lightManager: LightManager by inject()

    private val settingsRepository: SettingsRepository by inject()

    private val binding get() = _binding!!
    private var _binding: FragmentLightBinding? = null

    /**
     * Receive and handle commands from screen module
     */
    private val broadcastReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            // Get brightnessObservable value
            val brightness: Int = intent.getIntExtra(ScreenModule.EXTRA_BRIGHTNESS_PERCENT, 100)
            setBrightness(brightness)
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentLightBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.fab.keepScreenOn = settingsRepository.isKeepScreenOn
        binding.fab.setOnClickListener { lightManager.turnOff() }
    }

    override fun onResume() {
        super.onResume()
        LocalBroadcastManager.getInstance(
            requireContext()
        ).registerReceiver(
            broadcastReceiver, IntentFilter(ScreenModule.ACTION_SCREEN_MODULE)
        )
    }

    override fun onPause() {
        super.onPause()

        LocalBroadcastManager.getInstance(requireContext()).unregisterReceiver(broadcastReceiver)
    }

    override fun onStop() {
        super.onStop()

        // Turn off light because screen not visible and for this mode it makes no sense
        if (lightManager.isTurnedOn) lightManager.turnOff()
    }

    private fun setBrightness(percent: Int) {
        val color = 255 * percent / 100
        binding.layoutRoot.setBackgroundColor(Color.rgb(color, color, color))
    }
}